src/test/java/com/google/cloud/spanner/pgadapter/JdbcSimpleModeTest.java

```diff
@@ -0,0 +1,361 @@
+
+        assertTrue(statement.getMoreResults());
+        try (ResultSet resultSet = statement.getResultSet()) {
+          assertTrue(resultSet.next());
+          assertEquals(2L, resultSet.getLong(1));
+          assertFalse(resultSet.next());
+        }
+
+        // getMoreResults() should now return false. We should also check getUpdateCount() as that
+        // method should return -1 to indicate that there is also no update count available.
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+      }
+    }
+
+    // The server will also receive a 'SELECT 1' statement from PgAdapter, as PgAdapter calls
+    // connection.isAlive() before using it.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(3, requests.size());
+    assertEquals(SELECT1.getSql(), requests.get(0).getSql());
+    assertEquals(UPDATE_STATEMENT.getSql(), requests.get(1).getSql());
+    assertEquals(SELECT2.getSql(), requests.get(2).getSql());
+  }
+
+  @Test
+  public void testBeginAndDml() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(createUrl())) {
+      try (java.sql.Statement statement = connection.createStatement()) {
+        // batch a BEGIN [TRANSACTION] statement together with an update statement.
+        assertFalse(statement.execute(String.format("BEGIN; %s;", UPDATE_STATEMENT)));
+        assertEquals(0, statement.getUpdateCount());
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(2, statement.getUpdateCount());
+
+        // getMoreResults() should now return false. We should also check getUpdateCount() as that
+        // method should return -1 to indicate that there is also no update count available.
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+
+        // Execute an insert statement as part of the transaction.
+        assertFalse(statement.execute(INSERT_STATEMENT.getSql()));
+        assertEquals(1, statement.getUpdateCount());
+
+        assertFalse(statement.getMoreResults());
+        assertEquals(-1, statement.getUpdateCount());
+
+        statement.execute("COMMIT");
+      }
+    }
+
+    // The server will also receive a 'SELECT 1' statement from PgAdapter, as PgAdapter calls
+    // connection.isAlive() before using it. This should happen before the transaction is started.
+    List<ExecuteSqlRequest> requests = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class);
+    assertEquals(3, requests.size());
+    assertEquals(SELECT1.getSql(), requests.get(0).getSql());
+    assertTrue(requests.get(0).getTransaction().hasSingleUse());
+
+    // The first statement in the transaction will include the BeginTransaction option.
+    assertEquals(UPDATE_STATEMENT.getSql(), requests.get(1).getSql());
+    assertTrue(requests.get(1).getTransaction().hasBegin());
```

Comment:
same here

Reply:
(See also my comment above regarding the `SELECT 1` statement, and remember that the `SELECT 1` statement is something that was executed internally by PGAdapter, and is not part of the batch that we are actually executing in this test)

This is a very good question, and I'll try to explain it using the specific example, as it is a little complex and requires some knowledge of how the Java client library and Connection API has been implemented.

This test executes the following series of statements.

```sql
statement.execute("BEGIN; UPDATE FOO SET BAR=1 WHERE BAZ=2;");
statement.execute("INSERT INTO FOO VALUES (1)");
statement.execute("COMMIT;");
```

The Java client library and the Connection API use what is called 'inlined `BeginTransaction`'. This means that the request to begin a read/write transaction will be inlined (included) with the first statement in the transaction, instead of being executed as a separate `BeginTransaction` RPC. So this means:
1. The `BEGIN;` statement is internally in the Connection API only translated to setting a flag that the application has started a transaction. It does __not__ execute any RPCs.
2. The `UPDATE FOO SET BAR=1 WHERE BAZ=2;` is the first statement in the read/write transaction. The request will therefore also carry an additional option to begin a new read/write transaction, and to use that transaction for the statement. The statement will return two things:
2.1. The ID of the transaction that was started. This is set as a property on the connection internally in the Connection API and is not returned to the client application.
2.2. The update count of the statement. This is returned to the client application.
3. The `INSERT INTO FOO VALUES (1)` will use the transaction ID that was returned by step 2.
4. The `COMMIT;` will commit the transaction that was started by step 2.

So:
1. `assertTrue(requests.get(1).getTransaction().hasBegin());` verifies that the `UPDATE` statement really included a `BeginTransaction` option in the request.
2. `assertTrue(requests.get(2).getTransaction().hasId());` verifies that the `INSERT` statement uses an existing transaction ID (the transaction ID that was returned by the `UPDATE` statement).